### PR TITLE
Use depth and index to lookup type metadata artificial variables

### PR DESCRIPTION
**Pull request by adrian-prantl**
Use depth and index to lookup type metadata artificial variables  …
instead of using name and decl context.

The advantages of this approach are three-fold:
- This is necessary to support inlined generic functions.
- We can retire the debugger-specific type name manfgling mode for archetypes.
- This saves 270kb of debug information in the x86_64 libSwiftCore.dylib alone.

<rdar://problem/38306256>

### CODE REVIEW

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/swift-lldb/pull/671
@swift-ci test

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/swift-lldb/pull/671
@swift-ci test

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/swift-lldb/pull/671
@swift-ci test

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/swift-lldb/pull/671
@swift-ci test

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/swift-lldb/pull/671
@swift-ci test

File: lib/IDE/TypeReconstruction.cpp [hunk 1]
```diff
@@ -1005,6 +1005,28 @@ static void VisitNodeDestructor(
   }
 }
 
+static void VisitNodeDependentMember(ASTContext *ast,
```
Comment (dcci) [inline lib/IDE/TypeReconstruction.cpp]: We should really add a `lldb-moduleimport-test` test for this.

File: lib/IRGen/DebugTypeInfo.cpp [hunk 2]
```diff
@@ -56,6 +56,7 @@ DebugTypeInfo DebugTypeInfo::getFromTypeInfo(DeclContext *DC,
     // encounter one.
     size = Size(0);
   }
+
```
Comment (dcci) [inline lib/IRGen/DebugTypeInfo.cpp]: Spurious newline I assume?

**Review by dcci — CHANGES_REQUESTED** [review]
Some comments inline. I'll try to review more carefully in the next hours.

**Comment by adrian-prantl** [comment]
@dcci: Can you take another look?

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/swift-lldb/pull/671
@swift-ci test

**Review by dcci — APPROVED** [review]
LGTM

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/swift-lldb/pull/671
@swift-ci test platform Linux

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/swift-lldb/pull/671
@swift-ci test

**Pull request merged**